runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -368,6 +375,34 @@ static iree_status_t iree_hal_hip_device_initialize_internal(
     }
   }
 
+  if (iree_status_is_ok(status)) {
+    for (iree_host_size_t i = 0; i < device->device_count; ++i) {
+      iree_hal_buffer_params_t buffer_params = {
+          .usage = IREE_HAL_BUFFER_USAGE_TRANSFER |
+                   IREE_HAL_BUFFER_USAGE_MAPPING_SCOPED,
+          .access = IREE_HAL_MEMORY_ACCESS_READ | IREE_HAL_MEMORY_ACCESS_WRITE |
+                    IREE_HAL_MEMORY_ACCESS_DISCARD,
+          .type = IREE_HAL_MEMORY_TYPE_HOST_VISIBLE |
+                  IREE_HAL_MEMORY_TYPE_DEVICE_VISIBLE,
+          .queue_affinity = (iree_hal_queue_affinity_t)1 << i,
+          .min_alignment = 0,
+      };
+      status = iree_hal_allocator_allocate_buffer(
+          device->device_allocator, buffer_params,
+          params->file_transfer_buffer_size,
+          &device->devices[i].file_transfer_staging_buffer);
+      if (!iree_status_is_ok(status)) {
+        break;
+      }
+      device->devices[i].file_transfer_staging_buffer_head = 0;
+      device->devices[i].file_transfer_staging_buffer_tail = 0;
+      iree_slim_mutex_initialize(
```

Comment:
just fyi note that this is only safe because we calloc - if you are ever doing this into reused/uninitialized memory then you'll want to ensure that there's no code path that can break early from this loop and call iree_notification_deinitialize later on (such as if the alloc above fails, this breaks early, then destroy() is run to clean up and walk the entire list)

Reply:
Yes, I was aware there. But it's definitely good to call it out.